cpp/src/arrow/io/file.cc

```diff
@@ -421,7 +370,21 @@ class OSFile {
   FileMode::type mode() const { return mode_; }
 
  protected:
-  PlatformFilename path_;
+  Status SetFileName(const std::string& file_name) {
+    try {
+#if defined(_MSC_VER)
+      std::codecvt_utf8_utf16<wchar_t> utf16_converter;
+      file_name_.assign(file_name, utf16_converter);
```

Comment:
Explicit conversion type is still required here for correct translation of utf-8 symbols into utf-16.